### ISSUE TITLE
🔌 Plugin: Events UI

### ISSUE THREAD

**Issue by awanlin** [backstage/backstage#20091]
### 🔖 Summary

It would be really helpful to have a UI for the Events framework to be able to view events that Backstage has consumed, their status, and potentially replay them if needed.

Viewing events:

![Screenshot 2023-09-21 at 12 18 49 PM](https://github.com/backstage/backstage/assets/67169551/14a95ee4-cc8b-4b0e-9b39-68718ef6f72e)

Event payload:

![Screenshot 2023-09-21 at 12 19 18 PM](https://github.com/backstage/backstage/assets/67169551/6aefa76f-2bc5-462f-bc00-cbd188eff19b)

Event metadata:

![Screenshot 2023-09-21 at 12 19 32 PM](https://github.com/backstage/backstage/assets/67169551/a8da907e-9b6b-488b-b28c-82bf18933dd8)

### ✌️ Context

At Keyloop we have various sources sending in events - Azure DevOps, Bitbucket, and GitHub - and being able to see the events in one place has been extremely handy. We use this "dashboard" in the DevTools plugin along with what we call a `DatabaseEventBroker` that stores all the events. There is also a process to truncate this data keeping a configured amount of events.

### 👀 Have you spent some time to check if this plugin request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, I don't have time to work on this right now

**Comment by bforbis** [comment]
@awanlin this looks like it potentially could involve a lot of writes to the DB if the events are coming in fast enough. Maybe a toggle that lets you turn on/off recording mode might be useful to have. I've used some systems with this behavior where recording mode will only capture the next X amount of events, then it goes back to the default mode of not persisting them.

**Comment by awanlin** [comment]
Yeah, for sure, we haven't had any issues like that but we are on the smaller side and doesn't mean they can't happen.

**Comment by awanlin** [comment]
One feature we don't have but would like is a way to bundle the events together. There's no "id" on the payload but if there was we could link them up

**Comment by vinzscam** [comment]
This is great. We also need to mention that in some cases we might want to avoid showing certain events or data within events, as they might contain confidential data

**Comment by pjungermann** [comment]
As a prerequisite, we **might** want to have a different `EventsService` implementation. E.g., we were thinking about having a DB/Postgre-based implementation instead of the in-memory implementation (or as an alternative). The current implementation has some downsides like events can only be passed within the same instance (no separation of backends possible). Such an implementation might only be usable until a certain size/volume, too, might need to be replaced by an external system like Kafka at some point.

Another option could be something like an on-demand event stream. You could connect from the frontend as a subscriber on the backend and receive events as they get received.
However, as events are only processed and passed by one instance of the cluster, you would not get all until you subscribe at each instance.

A combination of both might also make sense, depending on how events would get persisted at a potential DB-based solution or others.

**Comment by pjungermann** [comment]
Status can be tricky, too. Each subscriber would have their own status.

**Comment by Rugvip** [comment]
Tbh I find this to be roughly the equivalent of storing and surfacing all HTTP requests that been received by the backend - it's a bit too much. +1 to being able to temporarily tap into the event stream so see what's coming through though 👍

**Comment by awanlin** [comment]
@pjungermann I don't have access to the `DatabaseEventBroker` implementation I mentioned here but I wrote it and remember the design at a high level. Probably not a pattern to follow, lol. We had an `event-broker-backend` which had a simple API with a router that would take the event and save it to its database. Then `DatabaseEventBroker` would just pass on the event to that API route. The frontend was all in an `event-broker` plugin which then got added to DevTools.

@Rugvip for sure, as I mentioned before this worked fine in the case with a small to mid-size org but I'm sure it needs a better design to handle load properly for larger orgs. My experience with this was that it was really handy when initially setting up events to be able to see the data but also debugging when there was issues. Often you can't easily trigger the event again so having the data was really helpful so only having it on after the fact might reduce how helpful this is. I'm sure there's probably a way to make this all work though 👍
